File: 0515) Find Largest Value in Each Tree Row.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        int level = 0;
        dfs(root, level, ans);
        return ans;
    }
    public void dfs(TreeNode root, int level, List<Integer> ans)
    {
        if(root == null)
        {
            return;
        }
        if(level == ans.size())
        {
            ans.add(root.val);
        }
        else
        {
            ans.set(level, Math.max(ans.get(level), root.val));
        }
        dfs(root.left, level + 1, ans);
        dfs(root.right, level + 1, ans);
    }
}

File: 0515) Find Largest Value in Each Tree Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> ans;
        int level = 0;
        helper(root, level, ans);
        return ans;
    }
    void helper(TreeNode *root, int level, vector<int> &ans)
    {
        if(root == nullptr)
        {
            return;
        }
        if(level == ans.size())
        {
            ans.push_back(root->val);
        }
        else
        {
            ans[level] = max(ans[level], root->val);
        }
        helper(root->left, level+1, ans);
        helper(root->right, level+1, ans);
    }
};
